twister: support standalone run without build

hakehuang opened the pull request:
Enable twister to support standalone run with prebuild image.
skip west local dependency on build information from cmake.

e.g.
```
scripts/twister -p frdm_k64f --device-testing \
--hardware-map ~/frdm_k64f.yml --test-only \
--load-tests testplan.json \
--west-flash="--device=MK64FN1M0xxx12,--bin-file=/home/shared/temp/zephyr.bin" \
--force-board-config-path=~/zephyr_runner/
```
Read the twister guide "Running test only mode" for more

This will enable twister to support distribute build and run

hakehuang committed:
west: add force option to run from command line

add -f force to flash command, so that we can flash pre-build image,
and with twister frawwork to verify the result
-d is to define the build_path which shall have the zephyr/.config
.config file can be generate from hello_work for given platform

e.g.

west flash -f --skip-rebuild --runner jlink \
-d=~/zephyr_runner/                   \
--tool-opt='-SelectEmuBySN 621000000' \
--  '--device=MK64FN1M0xxx12'         \
--bin-file ./zephyr.bin

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

hakehuang committed:
twister: enable run from prebuild binary

Add --force-build-board-path option
This will enable run with prebuild image

e.g.

scripts/twister -p frdm_k64f --device-testing \
--hardware-map ~/frdm_k64f.yml --test-only \
--load-tests testplan.json \
--west-flash="--device=MK64FN1M0xxx12,\
--bin-file=/home/shared/temp/zephyr.bin" \
--force-board-config-path=~/zephyr_runner/

Read the twister guide "Running test only mode" for more

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

hakehuang committed:
twister: document: update twister document

Add guider for how to run test only

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

enjiamai commented:
Hi , @hakehuang , the user scenario confuse me in two points:

(1) This method only supoort running one pre-built binary. 
(Actually we can use the existing -b then --test-only to run multiple tests.) 

(2) I think you also can do it this way to achieve you user scenario:
```  
twister --hardware-map ~/map.yml -b -T tests/kernel/semaphore/semaphore/

cp ./twister-out/...../zephyr.bin /home/shared/temp/zephyr.bin 
```
Assume you have following hardware map file:
``` 
- connected: true
  fixtures:
    - uart_loopback
  id: 024002010ED63E73F328C3CB
  platform: frdm_k64f
  product: MBED CMSIS-DAP
  runner: pyocd
  serial: /dev/ttyACM0
  runner_params:
    - --bin-file=/home/shared/temp/zephyr.bin
```
Then run it:
```
twister --device-testing --hardware-map ./map.yml --test-only -vv
```

So I have some concern of adding such parameters and its user scenario, could you please help to clarify?

mbolivar-nordic commented on scripts/west_commands/run_common.py:
```diff
@@ -127,6 +127,8 @@ def add_parser_common(command, parser_adder=None, parser=None):
         etc.'''))
     group.add_argument('-H', '--context', action='store_true',
                        help='print runner- and build-specific help')
+    group.add_argument('-f', '--force', action='store_true',
+                       help='force download with command line ignore load error')
```
Can you please refactor this routine so that only `flash` has this option? It looks like `debug` and `debugserver` do not support this, so I don't think they should have the option.

hakehuang replied:
good point, thanks, refactor.

nashif commented on doc/develop/test/twister.rst:
```diff
@@ -759,3 +759,35 @@ dependencies between test cases.  For native_posix platforms, you can provide
 argument to twister. See :ref:`Shuffling Test Sequence <ztest_shuffle>` for more
 details.
+
+Running test only mode
```
Not a big fan of this approach. We should be able to support building on one host and running on another with less manual steps, supporting multiple boards and without too many modifications to flashers and runners...
IMO you are losing the connection between the original artefacts and the run results and moving prebuilt binaries  outside of the environment to load them somewhere else will make traceability and management of such environments very difficult.

Is it only about not copying everything produced in the build step? How about remove all build artefacts and just leave those needed for flashing and running? How are you going to support dealing with multiple boards in your solution and so on...?

hakehuang replied:
@nashif , you are right this approach is trying to decouple the build environments and running target. but leave user to choose to keep the traceabilities. .e.g they can wrapped remote build and remote run together in jenkins scripts.

> Is it only about not copying everything produced in the build step? How about remove all build artefacts and just leave those needed for flashing and running? How are you going to support dealing with multiple boards in your solution and so on...?

The current build artifect are linked with case name,  but actually what we need to run is only the flash /debugger  information. this can be once for all effort, instead of copy a build folder and delete many other contents.

To support multiply boards, we need use another process to do so. It is not a all in one feature, this feature is provide such as when we have a powful machine to build all binaries, and send each binary to runner, and runner can execute them one by one on given board. So we will need a jenkins like system to wrapper the steps and trigger different testplan for each application on each board.

hakehuang replied:
> another with less manual steps, supporting multiple boards and without too many modifications to flashers and runners...

the manual step lst here are one time job when you setting a board in run machine. and the modifications to flash and runners is just to remove the hard check on supportting files which dynamically linked with testcase name. instead using a static path can do the work, given the flash size and address is fixed for given board.

PerMac replied:
I agree with @nashif. Looks quite fiddly. We already have `--runtime-artifact-cleanup` to leave only a bare minimum of build artifacts. However, I found it is not working if such artifacts are to be used later on with --test-only. I created a PR https://github.com/zephyrproject-rtos/zephyr/pull/46782  which expands the list of leftovers by a minimal set required to make flashing to work. Maybe such approach will fulfill your use case @hakehuang?

hakehuang replied:
> IMO you are losing the connection between the original artefacts and the run results and moving prebuilt binaries outside of the environment to load them somewhere else will make traceability and management of such environments very difficult.

@nashif , I will show the demo today, in test plan there is a runid, which is abided with the binary, so the tracebility is there.

hakehuang committed:
twister: do not quit twister when build-only

--save-tests will quit after execute. But as run-id add
we need have test plan has aligned run-id with build image,
so do not quit twister if has build-only and save-tests together

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
